File: app/src/main/java/com/example/utils/PatientSharing.kt
package com.example.utils

import com.example.data.PatientEntity
import org.json.JSONArray
import org.json.JSONObject

object PatientSharing {
    private fun patientToJs(patient: PatientEntity): JSONObject {
        return JSONObject().apply {
            put("name", patient.name)
            put("abo", patient.abo)
            put("diagnosis", patient.diagnosis)
            put("lmp", patient.lmp)
            put("edd", patient.edd)
            put("us", patient.us)
            put("hb", patient.hb)
            put("hct", patient.hct)
            put("plt", patient.plt)
            put("wbc", patient.wbc)
            put("urea", patient.urea)
            put("creat", patient.creat)
            put("sgot", patient.sgot)
            put("sgpt", patient.sgpt)
            put("inr", patient.inr)
            put("rbs", patient.rbs)
            put("na", patient.na)
            put("k", patient.k)
            put("alb", patient.alb)
            put("acetone", patient.acetone)
            put("sugar", patient.sugar)
            put("pus", patient.pus)
            put("notes", patient.notes)
            put("labRecordsJson", patient.labRecordsJson)
        }
    }

    private fun jsToPatient(obj: JSONObject): PatientEntity {
        return PatientEntity(
            name = obj.optString("name", ""),
            abo = obj.optString("abo", ""),
            diagnosis = obj.optString("diagnosis", ""),
            lmp = obj.optString("lmp", ""),
            edd = obj.optString("edd", ""),
            us = obj.optString("us", ""),
            hb = obj.optString("hb", ""),
            hct = obj.optString("hct", ""),
            plt = obj.optString("plt", ""),
            wbc = obj.optString("wbc", ""),
            urea = obj.optString("urea", ""),
            creat = obj.optString("creat", ""),
            sgot = obj.optString("sgot", ""),
            sgpt = obj.optString("sgpt", ""),
            inr = obj.optString("inr", ""),
            rbs = obj.optString("rbs", ""),
            na = obj.optString("na", ""),
            k = obj.optString("k", ""),
            alb = obj.optString("alb", ""),
            acetone = obj.optString("acetone", ""),
            sugar = obj.optString("sugar", ""),
            pus = obj.optString("pus", ""),
            notes = obj.optString("notes", ""),
            labRecordsJson = obj.optString("labRecordsJson", "[]"),
            isSelectedForPrint = true
        )
    }

    fun serialize(patient: PatientEntity): String {
        return try {
            val obj = patientToJs(patient).apply {
                put("type", "FASTER_OBS_PATIENT")
                put("version", 1)
            }
            obj.toString(2)
        } catch (e: Exception) {
            ""
        }
    }

    fun serializeList(patients: List<PatientEntity>): String {
        return try {
            val root = JSONObject().apply {
                put("type", "FASTER_OBS_PATIENTS_LIST")
                put("version", 1)
                val array = JSONArray()
                patients.forEach { patient ->
                    array.put(patientToJs(patient))
                }
                put("patients", array)
            }
            root.toString(2)
        } catch (e: Exception) {
            ""
        }
    }

    fun deserialize(jsonStr: String): PatientEntity? {
        if (jsonStr.isBlank()) return null
        return try {
            val trimmed = jsonStr.trim()
            val obj = if (trimmed.startsWith("{")) {
                JSONObject(trimmed)
            } else {
                return null
            }
            
            val type = obj.optString("type", "")
            if (type != "FASTER_OBS_PATIENT" && !obj.has("name") && type != "FASTER_OBS_PATIENTS_LIST") {
                return null
            }
            
            if (type == "FASTER_OBS_PATIENTS_LIST") {
                val array = obj.optJSONArray("patients")
                if (array != null && array.length() > 0) {
                    return jsToPatient(array.getJSONObject(0))
                }
                return null
            }

            jsToPatient(obj)
        } catch (e: Exception) {
            null
        }
    }

    fun deserializeList(jsonStr: String): List<PatientEntity>? {
        if (jsonStr.isBlank()) return null
        return try {
            val trimmed = jsonStr.trim()
            val obj = if (trimmed.startsWith("{")) {
                JSONObject(trimmed)
            } else {
                return null
            }
            
            val type = obj.optString("type", "")
            
            if (type == "FASTER_OBS_PATIENTS_LIST") {
                val array = obj.optJSONArray("patients") ?: return emptyList()
                val list = mutableListOf<PatientEntity>()
                for (i in 0 until array.length()) {
                    list.add(jsToPatient(array.getJSONObject(i)))
                }
                return list
            }
            
            if (type == "FASTER_OBS_PATIENT" || obj.has("name")) {
                val p = jsToPatient(obj)
                return listOf(p)
            }
            
            null
        } catch (e: Exception) {
            null
        }
    }
}

File: ios-app/PatientSharing.swift
import Foundation

public struct PatientSharing {
    
    private static func patientToDict(_ patient: Patient) -> [String: Any] {
        return [
            "name": patient.name,
            "abo": patient.abo,
            "diagnosis": patient.diagnosis,
            "lmp": patient.lmp,
            "edd": patient.edd,
            "us": patient.us,
            "hb": patient.hb,
            "hct": patient.hct,
            "plt": patient.plt,
            "wbc": patient.wbc,
            "urea": patient.urea,
            "creat": patient.creat,
            "sgot": patient.sgot,
            "sgpt": patient.sgpt,
            "inr": patient.inr,
            "rbs": patient.rbs,
            "na": patient.na,
            "k": patient.k,
            "alb": patient.alb,
            "acetone": patient.acetone,
            "sugar": patient.sugar,
            "pus": patient.pus,
            "notes": patient.notes,
            "labRecordsJson": patient.labRecordsJson
        ]
    }
    
    private static func dictToPatient(_ dict: [String: Any]) -> Patient {
        return Patient(
            id: Int.random(in: 1000...999999),
            postId: 0,
            name: dict["name"] as? String ?? "",
            abo: dict["abo"] as? String ?? "",
            diagnosis: dict["diagnosis"] as? String ?? "",
            lmp: dict["lmp"] as? String ?? "",
            edd: dict["edd"] as? String ?? "",
            us: dict["us"] as? String ?? "",
            hb: dict["hb"] as? String ?? "",
            hct: dict["hct"] as? String ?? "",
            plt: dict["plt"] as? String ?? "",
            wbc: dict["wbc"] as? String ?? "",
            urea: dict["urea"] as? String ?? "",
            creat: dict["creat"] as? String ?? "",
            sgot: dict["sgot"] as? String ?? "",
            sgpt: dict["sgpt"] as? String ?? "",
            inr: dict["inr"] as? String ?? "",
            rbs: dict["rbs"] as? String ?? "",
            na: dict["na"] as? String ?? "",
            k: dict["k"] as? String ?? "",
            alb: dict["alb"] as? String ?? "",
            acetone: dict["acetone"] as? String ?? "",
            sugar: dict["sugar"] as? String ?? "",
            pus: dict["pus"] as? String ?? "",
            notes: dict["notes"] as? String ?? "",
            createdAt: Date().timeIntervalSince1970 * 1000,
            isSelectedForPrint: true,
            labRecordsJson: dict["labRecordsJson"] as? String ?? "[]"
        )
    }
    
    public static func serialize(_ patient: Patient) -> String {
        var dict = patientToDict(patient)
        dict["type"] = "FASTER_OBS_PATIENT"
        dict["version"] = 1
        
        do {
            let data = try JSONSerialization.data(withJSONObject: dict, options: [.prettyPrinted])
            return String(data: data, encoding: .utf8) ?? ""
        } catch {
            return ""
        }
    }
    
    public static func serializeList(_ patients: [Patient]) -> String {
        let listDicts = patients.map { patientToDict($0) }
        let root: [String: Any] = [
            "type": "FASTER_OBS_PATIENTS_LIST",
            "version": 1,
            "patients": listDicts
        ]
        
        do {
            let data = try JSONSerialization.data(withJSONObject: root, options: [.prettyPrinted])
            return String(data: data, encoding: .utf8) ?? ""
        } catch {
            return ""
        }
    }
    
    public static func deserializeList(_ jsonStr: String) -> [Patient]? {
        guard let data = jsonStr.trimmingCharacters(in: .whitespacesAndNewlines).data(using: .utf8) else {
            return nil
        }
        
        do {
            guard let json = try JSONSerialization.jsonObject(with: data, options: []) as? [String: Any] else {
                return nil
            }
            
            let type = json["type"] as? String ?? ""
            
            if type == "FASTER_OBS_PATIENTS_LIST" {
                guard let patientsArray = json["patients"] as? [[String: Any]] else {
                    return []
                }
                return patientsArray.map { dictToPatient($0) }
            }
            
            if type == "FASTER_OBS_PATIENT" || json["name"] != nil {
                let p = dictToPatient(json)
                return [p]
            }
            
            return nil
        } catch {
            return nil
        }
    }
}
